### PR TITLE
slider add properties:thumbImage,progressColor,progressBackgroundColor,thumbColor

### PR DESCRIPTION
**Pull request by linjson**
date:7.8->
1.add "splitTrack", "thumbOffset" properties in android platform
2."trackImage" is also working in android 

date:5.19->
just like iOS's properties: 
progressColor--> maximumTrackTintColor
progressBackgroundColor--> minimumTrackTintColor

"thumbImage" is also working in android 
"thumbColor" only in android

but these properties is supported android 6.0+

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @mkonicek and @sebmarkbage to be potential reviewers.

**Comment by facebook-github-bot-2** [comment]
@linjson updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@linjson updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@linjson updated the pull request.

**Comment by mkonicek** [comment]
Thanks for working on this @linjson. Can you please update the description? See https://github.com/facebook/react-native/blob/master/PULL_REQUEST_TEMPLATE.md

**Comment by mkonicek** [comment]
Adding the 'needs revision' label because we need an explanation of what this does as well as a test plan to be able to review this.

**Comment by yuriteixeira** [comment]
Eager to see this coming out. Nice job!

**Comment by yuriteixeira** [comment]
If you need help to test this change, just ping me. Desperately need this :)

**Comment by facebook-github-bot-8** [comment]
@linjson updated the pull request.

**Comment by linjson** [comment]
@yuriteixeira,I tried some times,but,ci/circleci test is not passed
How can I fix these bugs

File: Libraries/Components/Slider/Slider.js [hunk 1]
```diff
@@ -127,6 +127,26 @@ var Slider = React.createClass({
      * Used to locate this view in UI automation tests.
      */
     testID: PropTypes.string,
+
+    /**
+     * The color used for the progress
+     * @platform android >=21
```
Comment (mkonicek) [inline Libraries/Components/Slider/Slider.js]: I'm not sure how this will render on the website. Can you try using `cd website; npm install; npm start` and post a screenshot please? It might not render well, in that case it's fine to explain this in the doc block above.

File: Libraries/Components/Slider/Slider.js [hunk 2]
```diff
@@ -127,6 +127,26 @@ var Slider = React.createClass({
...
+     */
+    progressColor: PropTypes.string,
+
+    /**
+     * The color used for the progressBackgroundColor
+     * @platform android >=21
```
Comment (mkonicek) [inline Libraries/Components/Slider/Slider.js]: Same here.

File: Libraries/Components/Slider/Slider.js [hunk 3]
```diff
@@ -127,6 +127,26 @@ var Slider = React.createClass({
...
+     */
+    progressBackgroundColor: PropTypes.string,
+
+    /**
+     * The color used for the thumbColor
+     * @platform android >=21
```
Comment (mkonicek) [inline Libraries/Components/Slider/Slider.js]: Same here.

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 4]
```diff
@@ -1,36 +1,43 @@
 /**
  * Copyright (c) 2015-present, Facebook, Inc.
  * All rights reserved.
- *
+ * <p/>
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: Please remove, don't think we use `<p>` in comments across the codebase.

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 5]
```diff
@@ -1,36 +1,43 @@
...
 
 /**
  * Slider that behaves more like the iOS one, for consistency.
- *
+ * <p/>
  * On iOS, the value is 0..1. Android SeekBar only supports integer values.
  * For consistency, we pretend in JS that the value is 0..1 but set the
  * SeekBar value to 0..100.
- *
+ * <p/>
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: In theory the `<p>` are good here for rendering HTML docs but don't think we use them across the codebase - can you check? It's more important to stay consistent.

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 6]
```diff
@@ -1,36 +1,43 @@
...
  * during dragging).
  */
 public class ReactSlider extends SeekBar {
 
+  private static final String PROP_ICON_URI = "uri";
   /**
    * If step is 0 (unset) we default to this total number of steps.
    * Don't use 100 which leads to rounding errors (0.200000000001).
-   */ 
+   */
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: Thanks!

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 7]
```diff
@@ -108,4 +115,52 @@ private void updateValue() {
   private int getTotalSteps() {
     return (int) Math.ceil((mMaxValue - mMinValue) / mStep);
   }
+
+  public void setThumbImage(ReadableMap source) {
+    String uri = source != null ? source.getString(PROP_ICON_URI) : null;
+
+
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: remove empty lines please

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 8]
```diff
@@ -1,36 +1,43 @@
...
  * For consistency, we pretend in JS that the value is 0..1 but set the
  * SeekBar value to 0..100.
- *
+ * <p/>
  * Note that the slider is _not_ a controlled component (setValue isn't called
  * during dragging).
  */
 public class ReactSlider extends SeekBar {
 
+  private static final String PROP_ICON_URI = "uri";
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: Would something like "slider_icon_uri" work here? It's more descriptive. Or does iOS pass "uri"?

**Comment by mkonicek** [comment]
This is pretty cool! Can you add some screenshots from the UI Explorer to this pull request please?

**Comment by linjson** [comment]
example:
slider

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java (hunk 8 above)
Comment (linjson) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: using "require('./xxx.png')" or "require('image!xxx')"

File: Examples/UIExplorer/SliderExample.js [hunk 9]
```diff
@@ -167,4 +167,30 @@ exports.examples = [
       );
     }
   },
+  {
+    title: 'Custom thumb image',
```
Comment (janicduplessis) [inline Examples/UIExplorer/SliderExample.js]: There is already an iOS example for this prop we can just reuse it by removing the `platform: ios`.

File: Libraries/Components/Slider/Slider.js [hunk 10]
```diff
@@ -127,6 +127,32 @@ var Slider = React.createClass({
...
      */
     testID: PropTypes.string,
+
+    /**
+     * The color used for the progress
+     *
+     * *using in android >= 21*
+     * @platform android
+     */
+    progressColor: PropTypes.string,
```
Comment (janicduplessis) [inline Libraries/Components/Slider/Slider.js]: I think we should just use `maximumTrackTintColor` and `minimumTrackTintColor` props since it does the same thing as those ios props, this will make it easier to use for cross platform apps.

File: Libraries/Components/Slider/Slider.js [hunk 11]
```diff
@@ -158,9 +184,10 @@ var Slider = React.createClass({
     props.onSlidingComplete = onSlidingComplete && ((event: Event) => {
       onSlidingComplete && onSlidingComplete(event.nativeEvent.value);
     });
-
-    return <RCTSlider 
+    thumbImage=resolveAssetSource(thumbImage);
```
Comment (janicduplessis) [inline Libraries/Components/Slider/Slider.js]: Nit: Don't reassign here just move this directly in the prop declaration (`thumbImage={resolveAssetSource(...)}`)

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 12]
```diff
@@ -108,4 +129,76 @@ private void updateValue() {
   private int getTotalSteps() {
     return (int) Math.ceil((mMaxValue - mMinValue) / mStep);
   }
+
+  public void setThumbImage(ReadableMap source) {
+    String uri = source != null ? source.getString(PROP_ICON_URI) : null;
+    if (uri != null) {
+      if (uri.startsWith("http://") || uri.startsWith("https://")) {
```
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: Looking at some code that does something similar we also check for file:// (see [this](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/toolbar/ReactToolbar.java#L285))

The code to load drawables should probably be moved to some helper class as it seems to be copy pasted in a couple of places. -_- Let's leave this for another PR though.

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 13]
```diff
@@ -108,4 +129,76 @@ private void updateValue() {
...
+        setThumbFormUrl(uri);
+      } else {
+        Drawable thumd = getDrawableByName(uri);
```
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: `thumd` -> `thumb`

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java [hunk 14]
```diff
@@ -108,4 +129,76 @@ private void updateValue() {
...
+        setThumb(thumd);
+      }
+    }
+  }
+
+  private void setThumbFormUrl(final String uri) {
```
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: Typo: `setThumbFormUrl` -> `setThumbFromUrl`

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSliderManager.java [hunk 15]
```diff
@@ -148,6 +150,28 @@ public void setStep(ReactSlider view, double value) {
     view.setStep(value);
   }
 
+  @ReactProp(name = "thumbImage")
+  public void setThumbImage(ReactSlider view,ReadableMap value) {
```
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSliderManager.java]: Nit: Missing space after `,` in a few places in this file.

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSliderManager.java [hunk 16]
```diff
@@ -148,6 +150,28 @@ public void setStep(ReactSlider view, double value) {
...
+  @ReactProp(name = "progressBackgroundColor",customType = "Color")
+  public void setProgressBackgroundColor(ReactSlider view,Integer value) {
+    view.setProgressBackgroundColor(value);
+  }
+
+  @ReactProp(name = "thumbColor",customType = "Color")
+  public void setThumbColor(ReactSlider view,Integer value) {
+    view.setThumbColor(value);
+  }
+
```
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSliderManager.java]: Nit: remove these extra line breaks.

**Comment by janicduplessis** [comment]
Mostly looks good to me, pointed out a few issues. The main thing we should do is use the same prop as ios for the bar color.

File: Libraries/Components/Slider/Slider.js (hunk 11 above)
Comment (linjson) [inline Libraries/Components/Slider/Slider.js]: I has tested on iOS

**Comment by linjson** [comment]
thank you

**Comment by linjson** [comment]
slider

File: ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java (hunk 5 above)
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/views/slider/ReactSlider.java]: PTAL @linjson

**Comment by ericvicenti** [comment]
It seems that you're making a lot of changes in this PR. Can you try splitting them up and writing a clearer description of what you're changing and why?

I can see three specific changes that you're working on here: 
- Setting of Thumb Color
- new splitTrack boolean
- Setting thumbImage and trackImage

Closing this for now- please submit new PRs for these individual changes!
